lib/selector.js

```diff
@@ -3,10 +3,16 @@ let Prefixer = require('./prefixer')
 let Browsers = require('./browsers')
 let utils = require('./utils')
 
+const SELECTORS_CACHE = {}
+
 class Selector extends Prefixer {
   constructor (name, prefixes, all) {
     super(name, prefixes, all)
     this.regexpCache = {}
+    SELECTORS_CACHE[name] = {
```

Comment:
How we will lean this cache? Will we have a problem if PostCSS will process many very different files (or even different versions of the same files) during the same Node.js session (so the cache will be alive?)

If you want a per-document cache, you can use `node.root().autoprefixerSelectors` cache. It will be attached only for the same CSS root.

Reply:
Oops, that's a really good question. I need to say I didn't even thought about such cases. I'll try to check them asap. Thanks!

Reply:
I haven't tested it yet. I just thought that ```SELECTORS_CACHE``` is filled with data during the file/files preprocessing. So probably it should work fine with multiple files. But I'm not 100% sure. By the way the naming ```SELECTORS_CACHE``` is kinda misleading. It's not a cache, it's just a collection.

Reply:
Can we replace it to `this.selectors` since we use only instance methods and data there?

Reply:
Nope, it wouldn't work in such implementation. The idea with this global const is to keep all possible selectors with all relevant prefixes. And with ```this.selectors``` we only get relevant selectors for ```this.name```. But rule might contain several selectors.

I like your propose about using ```node.root().autoprefixerSelectors```. I haven't checked it yet and actually I don't know how it works but may be It could help to get rid of ```SELECTORS_CACHE```.

Reply:
I am afraid of putting an instance method to the global cache. If this method is not related to instance, we can move it to global function with Memoization cache.

Reply:
>I am afraid of putting an instance method to the global cache.

Yep, I don't like it either.
>If this method is not related to instance, we can move it to global function with Memoization cache.

Could you please explain a bit more about that? Probably I don't get your point completely. Or maybe I just need to sleep a bit more...

In any case thanks for your time and thoughts about this PR!

Reply:
> Could you please explain a bit more about that?

Memoization is a pattern like this

```js
let cache = { }
function method (arg) {
  return … // Do heavy stuff
}
function methodCached (arg) {
  if (!cache[args]) {
    cache[args] = method(arg)
  }
  return cache[args]
}
```

It means if you have a pure function (which returns the same result on the same argument) you can cache the result in the value.

I thought that this pattern can be useful here, but found that I may understand the algorithm wrong (sorry, have too many things to do recently). Can you explain what do you do in `SELECTORS_CACHE`?

Reply:
Thank you for the explanation. I thought that your idea was to hide ```SELECTORS_CACHE``` in a closure. Glad I asked.

My idea is: the ```SELECTORS_CACHE``` is a list of all possible selectors, all possible related prefixes and related ```replace``` method of the instance. It's filled with data during the file preprocessing on any initialization of ```class Selector``` (and it's descendants). So after preprocessing we get smth like this:
```js
SELECTORS_CACHE = {
    '::placeholder': {
        possible: ['-moz-', '-webkit-'], // all possible prefixes for '::placeholder'
        replace, // consider it as a replace method of new Selector('::placeholder'),
    },
    ':read-only': {
        possible: ['-moz-'], // all possible prefixes for ':read-only'
        replace, // consider it as a replace method of new Selector(':read-only'),
    }
    // and so on
};
```
Why do I need to keep all that data in one place? Cause later when ```Autoprefixer``` adds prefixes there could be rule with several selectors, for example ```.a::placeholder, .b:read-only```, and while the rule is processed with, the ```::placeholder``` instance, we need access not only to ```::placeholder``` prefixes and ```replace``` method but to ```:read-only``` prefixes and ```replace``` method too.

Hope I described my idea clear enough. And probably there is another (more proper) way to get the same behavior but I don't know it 😃

Reply:
If `replace` doesn’t not depends on instance data, can we move it to pure function? `replaceSelector()`

If possible selectors don’t rely on Browserslist config (because it is all possible prefixes) can we move it to pure function too? `getSelectorPossiblePrefixes()`

Then we can add Memoization to these two functions and call them directly from `lib/proccessor` or `lib/selector`. It could make code easier. Right now it is too complex to understand it without looking deeper and witching between files (I am afrading that next developers will not understand it too).

Reply:
I can try to do so. My concern is that ```replace``` method uses the instance method ```this.prefixed(prefix)``` inside. But hacks ```::placeholder``` and ```:fullscreen``` are overriding it. The ```::placeholder``` also overrides ```this.possible()``` method.

Reply:
> My concern is that replace method uses the instance method this.prefixed(prefix) inside. 

If it really rely on instance, you can't detouch it from this instance.

Don't forget that PostCSS can be called on many different CSS files.

Reply:
> If it really rely on instance, you can't detouch it from this instance.

That's make sense.
I'll try to implement your idea.

> Don't forget that PostCSS can be called on many different CSS files.

Can it process different files with different browser settings?

Reply:
> Can it process different files with different browser settings?

Yeap. The same instance can be used for different target browsers or different whitespace settings.

It will be safer just to think that every PostCSS call must be independent from each other.

Reply:
I don't know what was I thinking when I decided to use global ```const SELECTORS_CACHE```. Each instance of class ```Selector``` gets property ```all``` on its init. And ```all``` is just the ```Prefixes``` instance that already contains all the data we need to process grouping rules :smile: